### PR TITLE
Python script to register application for a workspace.

### PR DESCRIPTION
**Pull request by askew**
# PR for issue #211

The app registration in Azure AD required for authenticating a workspace will be created manually and then the app id passed to the create workspace API.

This PR has a Python script that can be used to create the app registration in Azure AD.

`workspace-app-reg.py -n <tre-name> -w <workspace-name>`

The option `--tre-name` or `-n` should be the name of the TRE, i.e. `"TRE"`
The option `--workspace-name` or `-w` should be the name of the workspace.

Both TRE name and workspace name are required.

The script uses the current Azure CLI credentials to call into Microsoft Graph. It will display the name of the tenant and the name of the signed in user and ask for confirmation. For example:
```
vscode ➜ /workspaces/tre (askew/task-211 ✗) $ ./scripts/workspace-app-reg.py -n TRE -w "Test Workspace"
You are about to create app registrations in the Azure AD Tenant "researchenv.onmicrosoft.com", signed in as "Stephen Askew"
Do you want to continue? (y/N) 
```

This prompt can be skipped using the option `--force` or  `-f`.

### CODE REVIEW

File: scripts/workspace-app-reg.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        resp = graph.post("/applications", json=apiApp, headers={'Content-Type': 'application/json'})
+        content = resp.json()
+        if resp.ok:
+            apiAppId = content['appId']
+            print(f"Created application \"{content['displayName']}\" (appid={apiAppId})")
+            graph.post(f"/applications/{content['id']}/owners/$ref")
+        else:
+            print(content)
+
+    if apiAppId is not None and len(apiAppId) != 0:
```
Comment (sachinkundu) [inline scripts/workspace-app-reg.py:188]: if not apiAppId

**Review by sachinkundu — APPROVED** [review]
consider the changes

File: scripts/workspace-app-reg.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
+#!/usr/local/bin/python3
+
+import sys, getopt, uuid
```
[the remark this answers is not in the document]
Comment (askew) [reply scripts/workspace-app-reg.py]: Great suggestion, that is much easier to use!

File: scripts/workspace-app-reg.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        self.code: str = error['error']['code']
+        self.message: str = error['error']['message']
+        self.innerError: dict = error['error']['innerError']
+
+
+class CliGraphClient(GraphClient):
+    def __init__(self):
+        super().__init__(credential=AzureCliCredential(), scopes=['https://graph.microsoft.com/.default'])
+
+    def get(self, url: str, **kwargs):
```
[the remark this answers is not in the document]
Comment (askew) [reply scripts/workspace-app-reg.py:25]: New version makes more use of the class.

File: scripts/workspace-app-reg.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    if len(sps) == 0:
+        graph.post("/servicePrincipals", json={
+            "appId": appId,
+            "appRoleAssignmentRequired": True,
+            "tags": ['WindowsAzureActiveDirectoryIntegratedApp'],
+        }, headers={'Content-Type': 'application/json'})
+
+
+def main():
+    global me
```
[the remark this answers is not in the document]
Comment (askew) [reply scripts/workspace-app-reg.py]: I've now removed all global variables.

File: scripts/workspace-app-reg.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        _double_check()
+
+    existing_app = get_existing_app(options.app_name)
+
+    # Generate new Guids if needed
+    ownerRoleId = str(uuid.uuid4())
+    researcherRoleId = str(uuid.uuid4())
+
+    # Get existing role and scope ids (in case we are updating the existing app)
+    if existing_app is not None:
```
[the remark this answers is not in the document]
Comment (askew) [reply scripts/workspace-app-reg.py]: Done

File: scripts/workspace-app-reg.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+    # Define the API application
+    apiApp = {
+        "displayName": options.app_name,
+        "appRoles": appRoles,
+        "signInAudience": "AzureADMyOrg"
+    }
+
+    if existing_app is not None:
+        apiAppId = existing_app['appId']
```
[the remark this answers is not in the document]
Comment (askew) [reply scripts/workspace-app-reg.py]: create and update now refactored into `CliGraphClient` methods.

File: scripts/workspace-app-reg.py (hunk 1 above)
Comment (askew) [reply scripts/workspace-app-reg.py]: This line has gone now, but useful to know the simpler form.

**Pull request merged**